src/Analyzer.JsonRuleEngine/Expressions/Expression.cs

```diff
@@ -34,59 +35,65 @@ internal abstract class Expression
+            EvaluateInternal(jsonScope, jsonLineNumberResolver, getResult: getResult, getEvaluation: null);
 
         /// <summary>
         /// Performs tasks common across <see cref="Expression"/> implementations, such as
         /// determining all paths to run against.
         /// </summary>
         /// <param name="jsonScope">The scope being evaluated.</param>
+        /// <param name="jsonLineNumberResolver">An <see cref="ILineNumberResolver"/> to
+        /// map JSON paths in the returned evaluation to the line number in the JSON evaluated.</param>
         /// <param name="getEvaluation">A delegate for logic specific to the child <see cref="Expression"/>, which
         /// generates a <see cref="JsonRuleEvaluation"/> for the specified <paramref name="jsonScope"/>.</param>
         /// <returns>A <see cref="JsonRuleEvaluation"/>, either populated with the evaluations generated from <paramref name="getEvaluation"/>,
         /// or the single <see cref="JsonRuleEvaluation"/> if only one was generated from <paramref name="getEvaluation"/>.</returns>
-        protected JsonRuleEvaluation EvaluateInternal(IJsonPathResolver jsonScope, Func<IJsonPathResolver, JsonRuleEvaluation> getEvaluation) =>
-            EvaluateInternal(jsonScope, getEvaluation: getEvaluation, getResult: null);
+        protected JsonRuleEvaluation EvaluateInternal(IJsonPathResolver jsonScope, ILineNumberResolver jsonLineNumberResolver, Func<IJsonPathResolver, JsonRuleEvaluation> getEvaluation) =>
+            EvaluateInternal(jsonScope, jsonLineNumberResolver, getEvaluation: getEvaluation, getResult: null);
 
         /// <summary>
         /// Performs tasks common across <see cref="Expression"/> implementations, such as
         /// determining all paths to run against.  Only <paramref name="getEvaluation"/>
         /// or <paramref name="getResult"/> should be populated, not both.
         /// </summary>
         /// <param name="jsonScope">The scope being evaluated.</param>
+        /// <param name="jsonLineNumberResolver">An <see cref="ILineNumberResolver"/> to
+        /// map JSON paths in the returned evaluation to the line number in the JSON evaluated.</param>
         /// <param name="getEvaluation">A delegate for logic specific to the child <see cref="Expression"/>, which
         /// generates a <see cref="JsonRuleEvaluation"/> for the specified <paramref name="jsonScope"/>.</param>
         /// <param name="getResult">A delegate for logic specific to the child <see cref="Expression"/>, which
         /// generates a <see cref="JsonRuleResult"/> for the specified <paramref name="jsonScope"/>.</param>
         /// <returns>The results of the evaluation.</returns>
         private JsonRuleEvaluation EvaluateInternal(
             IJsonPathResolver jsonScope,
+            ILineNumberResolver jsonLineNumberResolver,
```

Comment:
Oh this is weird. It's because it's used in the getEvaluation or getResult functions that are passed for classes that inherit from Expression... Is this a bad pattern or something? It's not intuitive when looking at this alone because VS just tells you that this parameter isn't referenced in the function so I feel like it can be something that someone can just remove. Maybe we make it optional with default value null instead? I'm not sure if this is any better though.

Reply:
Oh, I see what happened.  This is beautiful. :)

With the refactoring, passing `jsonLineNumberResolver` moved from the constructor to the `Evaluate` method of `Expression`s.  This meant I had to pass it into `EvaluateInternal` so it was available to pass to the *where* clause.

But then I decided to fix the issue of *where* clauses getting line numbers when they don't need to, and fixing that just meant passing it `null` and making it no longer required.  **But I missed that the *where* clause was the only thing that needed it!**  So now it's not even needed here, and can stay in the context of the concrete `Expression` instance. :)

Thanks for catching that!